src/Simulation/transition_decision_point.rs

```diff
@@ -0,0 +1,124 @@
+use crate::{
+    component::{State, Transition},
+    TransitionSystems::TransitionSystemPtr,
+};
+
+/// Represents a decision in a transition system: In the current `source` state there is a decision of using one of the `possible_decisions`.
+#[derive(Debug, Clone)]
+pub struct TransitionDecisionPoint {
+    source: State,
+    possible_decisions: Vec<Transition>,
+}
+
+impl TransitionDecisionPoint {
+    /// Constructs the initial [`TransitionDecisionPoint`] for a given [`TransitionSystemPtr`].
+    pub fn initial(system: &TransitionSystemPtr) -> Option<Self> {
+        system
+            .get_initial_state()
+            .map(|source| Self::from(system, &source))
+    }
+
+    /// Constructs the [`TransitionDecisionPoint`] from a `source` [`State`] and a given [`TransitionSystemPtr`].
+    pub fn from(system: &TransitionSystemPtr, source: &State) -> TransitionDecisionPoint {
+        let actions = system.get_actions();
+        let transitions: Vec<Transition> = actions
+            .into_iter()
+            // Map actions to transitions. An action can map to multiple actions thus flatten
+            .flat_map(|action| system.next_transitions_if_available(source.get_location(), &action))
+            // Filter transitions that can be used
+            .filter(|transition| transition.use_transition(&mut source.clone()))
```

Comment:
This is a bad api. You should not need a copy of source here. You should add an api that return whether a transition can be used, instead of making a copy and using the transition to check this.